### PR TITLE
Image Builder Improvements

### PR DESCRIPTION
**Pull request by ariefrahmansyah**
- refactor pyfunc base image

**What this PR does / why we need it**:

Image building is a mandatory process in the deployment of PyFunc-based Merlin components. The process abstracts the complexity of building and publishing a ready-to-be-used Docker image that contains the user's model source code, artifacts, and dependencies. However, there are some issues with the current image-building approach that lead to poor development experience and inefficient resource utilization. This PR addresses them by:

1. Refactoring Kaniko job specification generated by Merlin so that:
   a. it is more configurable -- we can set Kaniko's build arguments flags in the config and they will be propagated to the kaniko job spec
   b. it supports caching
2. Merlin API now generates a separate user's model dependencies files and upload it to GCS. This dependencies then will be installed separately during the image building (cause we use new Dockerfile)
   a. To achieve it, we introduce a new gsutil package to interface with GCS
3. Refactor base and application Dockerfiles used to build Docker images for pyfunc-server and batch prediction jobs
   a. Previously, the base Dockerfile had a step to install the pyfunc-server or batch prediction jobs dependencies. This step is now executed by the application Dockerfile
      i. Subsequently, this allows us to have only one base image instead of building base images for each Python version.
   b. New base Dockerfile will only install the platform components such as conda, spark, grpc library, but not the component itself
   c. New application Dockerfile will first download and install user's model dependencies; install component's dependencies; download model source code and artifacts; and finally dry-run the model

**Which issue(s) this PR fixes**:

Improves image building performance

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Checklist**

- [ ] Added unit test, integration, and/or e2e tests
- [ ] Tested locally
- [ ] Updated documentation
- [ ] Update Swagger spec if the PR introduce API changes
- [ ] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

**Comment by krithika369** [comment]
@ariefrahmansyah, not sure if this is WIP but could you please summarize the changes in the PR description? Like in https://github.com/caraml-dev/merlin/pull/483.

It will make it a lot easier to review. (Doesn't need to be elaborate, a summary of the motivation and the key changes will do.) Thanks!

File: api/pkg/imagebuilder/config.go [hunk 1]
```diff
@@ -26,6 +26,8 @@ type Config struct {
 	// Path to sub folder which is intended to build instead of using root folder
 	ContextSubPath string
 	// Dockerfile Path within the build context
+	BaseImage cfg.BaseImageConfig
```
Comment (krithika369) [inline api/pkg/imagebuilder/config.go:29]: Oh why do we have this config here and in `api/config.go` ? 🤔

**Review by krithika369 — COMMENTED** [review]
@ariefrahmansyah I'm still reviewing the PR (will get back to it shortly). Publishing the few comments I had first. 🙏🏽

File: .github/workflows/merlin.yml [hunk 2]
```diff
@@ -258,47 +260,46 @@ jobs:
           path: merlin.${{ needs.create-version.outputs.version }}.tar
           retention-days: ${{ env.ARTIFACT_RETENTION_DAYS }}
 
-  build-batch-predictor:
+  build-push-batch-predictor-base:
```
Comment (krithika369) [inline .github/workflows/merlin.yml:263]: What's the motivation for combining the build and publish steps? With this, we will be publishing the base images for every single commit to the PR / main, unlike previously, when the images are only pushed if the e2e succeeds. Shall we keep them separate as before?

File: .github/workflows/merlin.yml [hunk 3]
```diff
@@ -439,11 +440,9 @@ jobs:
     uses: ./.github/workflows/release.yml
     needs:
       - create-version
-      - e2e-test
+      # - e2e-test
```
Comment (krithika369) [inline .github/workflows/merlin.yml:443]: Is this intentional?

File: api/pkg/imagebuilder/imagebuilder.go [hunk 4]
```diff
@@ -129,6 +136,31 @@ func (c *imageBuilder) GetMainAppPath(version *models.Version) (string, error) {
...
+	condaEnv, err := c.gsutil.ReadFile(ctx, condaEnvUrl)
+	if err != nil {
+		return "", err
+	}
+
+	hash := sha256.New()
+	hash.Write([]byte(condaEnv))
+	hashEnv := hash.Sum(nil)
+
+	dependenciesUrl := fmt.Sprintf("gs://%s/merlin/dependencies/%x", artifactURL.Bucket, hashEnv)
```
Comment (leonlnj) [inline api/pkg/imagebuilder/imagebuilder.go:156]: `merlin/dependencies` sounds like it's pip freeze, `modeldependencies`?

File: api/pkg/gsutil/gsutil.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/bboughton/gcp-helpers/gsurl"
+	"github.com/pkg/errors"
+	"google.golang.org/api/option"
+)
+
+// parseURL exists to allow us to use 'url' as a request param
+var parseURL = url.Parse
```
[the remark this answers is not in the document]
Comment (ariefrahmansyah) [reply api/pkg/gsutil/gsutil.go:17]: Because func (g *gsutil) ParseURL(url string) also uses `url` with `string` data type as input. If we try to directly use url.Parse() in ParseURL(), we will get this error: `url.Parse undefined (type string has no field or method Parse)` because `url` here is the input variable `url`, not `net/url` package.

Other options are:

1. use alias import:

```
import (
    gourl "net/url"
)

...

func (g *gsutil) ParseURL(url string) (*URL, error) {
	u, err := gourl.Parse(url)
```

Or, 2. rename the input variable:

```
func (g *gsutil) ParseURL(gsURL string) (*URL, error) {
	u, err := url.Parse(gsURL)
```

I'll go with second option.

File: api/pkg/gsutil/gsutil.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	// Object is the name and or path of the object stored in the bucket. It
+	// should not start with a foward slash.
+	Object string
+}
+
+// Parse parses a Google Cloud Storage string into a URL struct. The expected
+// format of the string is gs://[bucket-name]/[object-path]. If the provided
+// URL is formatted incorrectly an error will be returned.
+func (g *gsutil) ParseURL(url string) (*URL, error) {
+	u, err := parseURL(url)
```
[the remark this answers is not in the document]
Comment (ariefrahmansyah) [reply api/pkg/gsutil/gsutil.go:85]: Discussed it here: https://github.com/caraml-dev/merlin/pull/500#discussion_r1418450859

File: api/pkg/imagebuilder/config.go (hunk 1 above)
Comment (ariefrahmansyah) [reply api/pkg/imagebuilder/config.go:29]: In cmd/setup.go we transform the config.ImageBuilderConfig to pkg.imagebuilder.Config: https://github.com/caraml-dev/merlin/blob/main/api/cmd/api/setup.go#L101-L141

When initializing imageBuilder object in pkg.imagebuilder (e.g. [NewModelServiceImageBuilder](https://github.com/caraml-dev/merlin/blob/main/api/pkg/imagebuilder/model_service_imagebuilder.go#L27)), we should be able to initialize it using its own internally-available Config struct, not external config.

File: python/batch-predictor/docker/app.Dockerfile [hunk 7]
```diff
@@ -13,19 +13,32 @@
...
+    else echo "GOOGLE_APPLICATION_CREDENTIALS is not set. Skipping gcloud auth command." > gcloud-auth.txt; \
+    fi
+
+# Download and install user model dependencies
+ARG MODEL_DEPENDENCIES_URL
+RUN gsutil cp ${MODEL_DEPENDENCIES_URL} conda.yaml
+RUN conda env create --name merlin-model --file conda.yaml
+
+# Copy and install batch predictor dependencies
+COPY --chown=${UID}:${GID} batch-predictor ${HOME}/merlin-spark-app
```
[the remark this answers is not in the document]
Comment (ariefrahmansyah) [reply python/batch-predictor/docker/app.Dockerfile:30]: 1. Will this command reuse the cache if the Kaniko build context doesn't change (i.e., 2 image building jobs performed within the same release using the same user conda env file)?
Yes. It will reuse the cache if no change in the build context.

2. If the build context changes (eg: 0.36.0 -> 0.37.0), is it command sure to get the new code (0.37.0) instead of reusing cache?
I haven't tested it, but in theory, it should get the new code and invalidate the cache. Will get back to you once I tested it.

File: .github/workflows/merlin.yml (hunk 2 above)
Comment (ariefrahmansyah) [reply .github/workflows/merlin.yml:263]: -> What's the motivation for combining the build and publish steps?
No strong opinion here. It was to fasten my development iteration. I'll separate it again.

**Comment by ariefrahmansyah** [comment]
-> Why is the python/sdk/test/pyfunc/model_2.joblib file removed in this PR? Isn't it needed by the integration tests?

When running pyfunc_integration_test.py from local machine, it will re-train the model hence updating the model. I deleted it  but forgot to restore the old one. I've restored it know.

File: api/pkg/imagebuilder/imagebuilder.go (hunk 4 above)
Comment (ariefrahmansyah) [reply api/pkg/imagebuilder/imagebuilder.go]: updated to model_dependencies

File: api/pkg/gsutil/gsutil.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	bytes, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read the content of %s", url)
+	}
+	return bytes, nil
+}
+
+// URL contains the information needed to identify the location of an object
+// located in Google Cloud Storage.
+type URL struct {
```
[the remark this answers is not in the document]
Comment (ariefrahmansyah) [reply api/pkg/gsutil/gsutil.go]: Updated.

File: api/pkg/imagebuilder/imagebuilder.go [hunk 9]
```diff
@@ -422,25 +461,23 @@ func (c *imageBuilder) createKanikoJobSpec(project mlp.Project, model *models.Mo
...
-	if !ok {
-		return nil, fmt.Errorf("No matching base image for tag %s", version.PythonVersion)
-	}
+	baseImageTag := c.config.BaseImage
 
 	kanikoArgs := []string{
 		fmt.Sprintf("--dockerfile=%s", baseImageTag.DockerfilePath),
 		fmt.Sprintf("--context=%s", baseImageTag.BuildContextURI),
-		fmt.Sprintf("--build-arg=MODEL_URL=%s/model", version.ArtifactURI),
 		fmt.Sprintf("--build-arg=BASE_IMAGE=%s", baseImageTag.ImageName),
```
[the remark this answers is not in the document]
Comment (ariefrahmansyah) [reply api/pkg/imagebuilder/imagebuilder.go]: Good idea. Updated!

File: api/pkg/imagebuilder/imagebuilder.go [hunk 10]
```diff
@@ -145,6 +177,13 @@ func (c *imageBuilder) BuildImage(ctx context.Context, project mlp.Project, mode
 	startTime := time.Now()
 	result := "failed"
 
+	// TODO(arief): create or get model dependency
```
[the remark this answers is not in the document]
Comment (ariefrahmansyah) [reply api/pkg/imagebuilder/imagebuilder.go]: -> we just need to check for the hashed file, whether it exists and create it if not, right?

The current implementation does not check whether the hashed file exists in GCS or not. It naively overwrites the hashed dependencies URL.

The comparison is:

Option 1: With checking will have at least 2 read operation + optional 1 write operation

1. Read user's raw model dependencies from GCS
2. Check hashed model dependencies URL exists in GCS or not
3. If exists, return hashed model dependencies URL
4. If not exists, write to hashed model dependencies URL in GCS

Option 2: Without checking will always have 1 read operation + 1 write operation

1. Read user's raw model dependencies from GCS
2. Write to hashed model dependencies URL in GCS

Option 2 is simpler but relatively more expensive (read operation in GCS is cheaper than write). I'm putting TODO to update it to Option 1 as originally discussed.

File: api/config/config.go [hunk 11]
```diff
@@ -203,15 +204,18 @@ type ImageBuilderConfig struct {
 	GcpProject                  string
 	ContextSubPath              string
 	DockerfilePath              string           `validate:"required" default:"./Dockerfile"`
+	BaseImage                   BaseImageConfig  `validate:"required"`
```
[the remark this answers is not in the document]
Comment (ariefrahmansyah) [reply api/config/config.go]: Yup, I've removed it.

File: api/pkg/imagebuilder/imagebuilder.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ariefrahmansyah) [reply api/pkg/imagebuilder/imagebuilder.go]: Yup, updated to Option 1: https://github.com/caraml-dev/merlin/blob/image-builder-improvements/api/pkg/imagebuilder/imagebuilder.go#L139

File: api/pkg/imagebuilder/imagebuilder.go [hunk 12]
```diff
@@ -100,35 +104,73 @@ const (
...
 		kubeClient:    kubeClient,
 		config:        config,
 		nameGenerator: nameGenerator,
+		gsutil:        gsutil,
 	}
 }
 
 // GetMainAppPath Returns the path to run the main.py of batch predictor, as configured via env var
 func (c *imageBuilder) GetMainAppPath(version *models.Version) (string, error) {
-	baseImageTag, ok := c.config.BaseImages[version.PythonVersion]
```
Comment (krithika369) [inline api/pkg/imagebuilder/imagebuilder.go:120]: I don't think we should remove this check entirely. Can we add a `SupportedPythonVersions []string` field under `ImageBuilderConfig` and use that to check? Ultimately, the pyfunc real time and batch server have some Python version constraints as does the SDK ([example](https://github.com/caraml-dev/merlin/blob/main/python/pyfunc-server/setup.py#L48)). We should capture this in the config and check here to prevent unnecessarily triggering image building for unsupported Python versions.

File: api/config/config.go [hunk 13]
```diff
@@ -202,17 +203,20 @@ type ClusterConfig struct {
 type ImageBuilderConfig struct {
 	ClusterName                 string `validate:"required"`
 	GcpProject                  string
+	GcsAuthenticationType       string
+	GcsCredentialsJson          string
```
Comment (krithika369) [inline api/config/config.go:207]: How is this value expected to be supplied during deployment time? If using Workload Identity, do we really need this?

Btw I just remembered we have this implementation, also used by Merlin as `MlflowDeleteService`, for model version delete:
* https://github.com/caraml-dev/mlp/blob/main/api/pkg/artifact/artifact.go#L7
* https://github.com/caraml-dev/mlp/blob/main/api/pkg/client/mlflow/mlflow.go

Can we make use of this instead of `gsutil.go` ? We will need to add the `ReadArtifact` and `WriteArtifact` APIs to the service [here](https://github.com/caraml-dev/mlp/blob/main/api/pkg/artifact/artifact.go#L11).

File: python/pyfunc-server/docker/Dockerfile [hunk 14]
```diff
@@ -15,19 +15,29 @@
...
-    conda env update --name merlin-model --file model/conda.yaml && \
-    pip install ${SDK_PATH} && \
-    python -m pyfuncserver --model_dir model --dry_run"
+# Download and dry-run user model artifacts and code
+ARG MODEL_ARTIFACTS_URL
+RUN gsutil -m cp -r ${MODEL_ARTIFACTS_URL} .
+RUN /bin/bash -c ". activate merlin-model && python -m pyfuncserver --model_dir model --dry_run"
 
-CMD ["/bin/bash", "./run.sh"]
+CMD ["/bin/bash", "/pyfunc-server/run.sh"]
```
Comment (krithika369) [inline python/pyfunc-server/docker/Dockerfile:43]: Since set `WORKDIR /pyfunc-server` above, is this change really needed?

File: python/batch-predictor/docker/app.Dockerfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (ariefrahmansyah) [reply python/batch-predictor/docker/app.Dockerfile:30]: Just tested it and the hypothesis is correct. The new code invalidates the cache.

File: api/pkg/imagebuilder/imagebuilder.go (hunk 12 above)
Comment (ariefrahmansyah) [reply api/pkg/imagebuilder/imagebuilder.go:120]: Oops, I forgot about SupportedPythonVersions. I've updated it

File: api/config/config.go (hunk 13 above)
Comment (ariefrahmansyah) [reply api/config/config.go]: I'm not aware of mlp.artifact package. Prior to writing this new gsutil package, I only see this gsutil being used by standard transformer -- which we should refactor too later. Turned out, this mlp.artifact is indirect dependencies of mlflow service.
I've updated the mlp.artifact package (thanks for the quick review there) and update merlin implementation to use it.

File: python/pyfunc-server/docker/Dockerfile (hunk 14 above)
Comment (ariefrahmansyah) [reply python/pyfunc-server/docker/Dockerfile:42]: It's not needed. During testing, I was playing around with and without workdir. And even though we have workdir there, I'm keeping /pyfunc-server in path to make it explicit.

File: .github/workflows/merlin.yml (hunk 3 above)
Comment (ariefrahmansyah) [reply .github/workflows/merlin.yml]: Uncommented it.

File: api/config/config.go [hunk 15]
```diff
@@ -99,8 +100,10 @@ type BaseImageConfig struct {
 	DockerfilePath string `validate:"required" json:"dockerfilePath"`
 	// GCS URL Containing build context
 	BuildContextURI string `validate:"required" json:"buildContextURI"`
+	// Path to sub folder which is intended to build instead of using root folder
+	BuildContextSubPath string `json:"buildContextSubPath"`
 	// path to main file to run application
-	MainAppPath string `validate:"required" json:"mainAppPath"`
+	MainAppPath string `json:"mainAppPath"`
```
Comment (krithika369) [inline api/config/config.go:106]: Why has the `validate:"required"` part been removed now? We're not somehow setting the default value elsewhere, so my understanding is we still need it. 🤔

File: api/pkg/imagebuilder/imagebuilder.go [hunk 16]
```diff
@@ -100,35 +104,91 @@ const (
...
+		kubeClient:      kubeClient,
+		config:          config,
+		nameGenerator:   nameGenerator,
+		artifactService: artifactService,
 	}
 }
 
+func (c *imageBuilder) validatePythonVersion(version *models.Version) error {
+	if version.PythonVersion == "" {
+		return fmt.Errorf("python version is not set model version %s", version.ID)
```
Comment (krithika369) [inline api/pkg/imagebuilder/imagebuilder.go:131]: Nit: `python version is not set for model version %s` ? The "for" is missing.

**Review by krithika369 — APPROVED** [review]
Left a couple of small comments. The rest LGTM. Thanks for this PR and the CI steps refactoring, @ariefrahmansyah !
